src/mon/ConfigKeyService.cc

```diff
@@ -34,61 +34,145 @@ static ostream& _prefix(std::ostream *_dout, const Monitor *mon,
 }
 
 const string ConfigKeyService::STORE_PREFIX = "mon_config_key";
+const string ConfigKeyService::META_PREFIX = "mon_config_key_meta";
+const string ConfigKeyService::META_KEY = "meta";
 
 void ConfigKeyService::refresh()
 {
-  dout(10) << __func__ << dendl;
-  config_keys.clear();
-
-  KeyValueDB::Iterator iter =
-    mon->store->get_iterator(STORE_PREFIX);
-  while (iter->valid()) {
-    config_keys[iter->key()] = iter->value();
-    iter->next();
+  version_t had_version = version;
+  bufferlist bl;
+  mon->store->get(META_PREFIX, META_KEY, bl);
+  list<string> changed_keys;
+  uint32_t total_keys = 0;
+  if (bl.length()) {
+    auto p = bl.begin();
+    ::decode(version, p);
+    while (!p.end()) {
+      string k;
+      ::decode(k, p);
+      changed_keys.push_back(k);
+      ::decode(total_keys, p);
+    }
+    dout(20) << __func__ << " meta shows version " << version << " and "
+	     << changed_keys.size() << " changed keys" << dendl;
+  } else {
+    version = 0;
+  }
+
+  if (had_version && version == had_version + 1) {
+    for (auto& k : changed_keys) {
+      bufferlist bl;
+      int r = mon->store->get(STORE_PREFIX, k, bl);
+      if (r == -ENOENT) {
+	config_keys.erase(k);
+      } else if (r == 0) {
+	config_keys[k] = bl;
+      } else {
+	derr << __func__ << " got " << cpp_strerror(r) << dendl;
+	ceph_abort();
+      }
+    }
+    dout(10) << __func__ << " refreshed v" << version << " with "
+	     << changed_keys.size() << " changed keys" << dendl;
+    if (total_keys == config_keys.size()) {
+      goto out;
+    }
+    derr << " have " << config_keys.size() << " expected " << total_keys
+	 << ", doing full reload" << dendl;
+  } else if (had_version && had_version == version) {
+    dout(20) << __func__ << " still v" << version << ", no change" << dendl;
+    if (total_keys == config_keys.size()) {
+      goto out;
+    }
+    derr << " have " << config_keys.size() << " but expected " << total_keys
+	 << ", doing full reload" << dendl;
   }
-  dout(10) << __func__ << " loaded " << config_keys.size() << " keys" << dendl;
+
+  {
+    config_keys.clear();
+    KeyValueDB::Iterator iter =
+      mon->store->get_iterator(STORE_PREFIX);
+    while (iter->valid()) {
+      config_keys[iter->key()] = iter->value();
+      iter->next();
+    }
+  }
+  dout(10) << __func__ << " full reload v" << version << " with "
+	   << config_keys.size() << " keys" << dendl;
+
+out:
+  metabl.clear();
+  ::encode(version + 1, metabl);
 }
```

Comment:
This looks a lot better than the initial patch introducing the cache, but I still worry of leaving the cache's size unbounded. Values should be small, but I fear how much this could impact if the user abuses the config-key service without having been warned first.